HealthVault/Classes/Caching/MHVThingCacheDatabase.m

```diff
@@ -0,0 +1,448 @@
+//
+//  MHVThingCacheDatabase.m
+//  MHVLib
+//
+// Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "MHVThingCacheDatabase.h"
+#import "MHVCommon.h"
+#import "MHVThingCacheDatabase+CoreDataModel.h"
+#import "MHVKeychainServiceProtocol.h"
+#import "MHVLogger.h"
+#import "EncryptedStore.h"
+#import "MHVThingTypes.h"
+#import "MHVThingCacheDatabase+CoreDataModel.h"
+#import "MHVCachedThing+Cache.h"
+#import "MHVCachedRecord+Cache.h"
+
+static NSString *kMHVCachePasswordKey = @"MHVCachePassword";
+
+@interface MHVThingCacheDatabase ()
+
+@property (nonatomic, strong) NSPersistentStoreCoordinator *persistentStoreCoordinator;
+@property (nonatomic, strong) NSManagedObjectModel *objectModel;
+@property (nonatomic, strong) NSManagedObjectContext *managedObjectContext;
+@property (nonatomic, strong) NSURL *databaseUrl;
+@property (nonatomic, strong) NSObject *lockObject;
+@property (nonatomic, assign) BOOL deleted;
+
+@property (nonatomic, strong) id<MHVKeychainServiceProtocol> keychainService;
+
+@end
+
+@implementation MHVThingCacheDatabase
+
+- (instancetype)initWithKeychainService:(id<MHVKeychainServiceProtocol>)keychainService
+{
+    self = [super init];
+    if (self)
+    {
+        _keychainService = keychainService;
+        _lockObject = [NSObject new];
+    }
+    return self;
+}
+
+- (void)deleteDatabase
+{
+    @synchronized (self.lockObject)
+    {
+        [[NSFileManager defaultManager] removeItemAtURL:self.databaseUrl error:nil];
+        
+        [self.keychainService setString:nil forKey:kMHVCachePasswordKey];
+        
+        _persistentStoreCoordinator = nil;
+        _objectModel = nil;
+        _managedObjectContext = nil;
+        _databaseUrl = nil;
+        
+        _deleted = YES;
+    }
+}
+
+#pragma mark - Create Objects
+
+- (MHVCachedThing *)newThingForRecord:(MHVCachedRecord *)record
+{
+    if (self.deleted)
```

Comment:
This isn't thread safe. It's possible that this method is called on a background thread, is paused after it makes it past the check of self.deleted, and the database then gets deleted.